src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -58,33 +58,46 @@ internal class SemanticTokensRangeHandler : IRequestHandler<LSP.SemanticTokensRa
         /// 
         /// Null when the client does not support sending refresh notifications.
         /// </summary>
-        private readonly AsyncBatchingWorkQueue? _semanticTokenRefreshQueue;
+        private readonly AsyncBatchingWorkQueue<Uri?>? _semanticTokenRefreshQueue;
 
         #endregion
 
         public SemanticTokensRangeHandler(
             IGlobalOptionService globalOptions,
             IAsynchronousOperationListenerProvider asynchronousOperationListenerProvider,
             LspWorkspaceRegistrationService lspWorkspaceRegistrationService,
+            LspWorkspaceManager lspWorkspaceManager,
             ILanguageServerNotificationManager notificationManager,
             ClientCapabilities clientCapabilities)
         {
             _globalOptions = globalOptions;
             _asyncListener = asynchronousOperationListenerProvider.GetListener(FeatureAttribute.Classification);
 
             _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
-
             _disposalTokenSource = new();
 
             if (clientCapabilities.Workspace?.SemanticTokens?.RefreshSupport is true)
             {
-                // Only send a refresh notification to the client every 0.5s (if needed) in order to avoid
+                // Only send a refresh notification to the client every 2s (if needed) in order to avoid
                 // sending too many notifications at once.  This ensures we batch up workspace notifications,
                 // but also means we send soon enough after a compilation-computation to not make the user wait
                 // an enormous amount of time.
-                _semanticTokenRefreshQueue = new AsyncBatchingWorkQueue(
-                    delay: TimeSpan.FromMilliseconds(500),
-                    processBatchAsync: c => notificationManager.SendNotificationAsync(Methods.WorkspaceSemanticTokensRefreshName, c),
+                _semanticTokenRefreshQueue = new AsyncBatchingWorkQueue<Uri?>(
+                    delay: TimeSpan.FromMilliseconds(2000),
+                    processBatchAsync: (documentUris, cancellationToken) =>
+                    {
+                        var trackedDocuments = lspWorkspaceManager.GetTrackedLspText();
+                        foreach (var documentUri in documentUris)
+                        {
+                            if (documentUri is null || !trackedDocuments.ContainsKey(documentUri))
+                            {
+                                return notificationManager.SendNotificationAsync(Methods.WorkspaceSemanticTokensRefreshName, cancellationToken);
+                            }
+                        }
+
+                        // LSP is already tracking all changed documents so we don't need to send a refresh request.
+                        return ValueTaskFactory.CompletedTask;
+                    },
                     asyncListener: _asyncListener,
                     _disposalTokenSource.Token);
```

Comment:
it would be good to use the IEqualityComparer overload of this and pass EqualityComparer<Uri?>.Default, that way we don't build up a list of tons of duplicate Uris in the queue.

Reply:
Great optimization suggestion, thanks!